File: app/src/main/java/com/ShaHar91/poemcollection/ui/base/BaseFragment.java
package com.ShaHar91.poemcollection.ui.base;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import com.ShaHar91.poemcollection.R;

import io.reactivex.disposables.CompositeDisposable;

public abstract class BaseFragment<VM extends BaseViewModel, C extends BaseComponent<VM>> extends Fragment {
    public static final String SHOW_BACK_ICON = "SHOW_BACK_ICON";

    protected VM viewModel;
    protected C component;

    protected abstract C createComponent();

    private CompositeDisposable compositeDisposable;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        compositeDisposable = new CompositeDisposable();
        component = createComponent();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onDestroy() {
        compositeDisposable.dispose();
        super.onDestroy();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        for (Fragment fragment : getChildFragmentManager().getFragments()) {
            fragment.onActivityResult(requestCode, resultCode, data);
        }
    }

    protected void configureToolbar(Toolbar toolbar, View toolbarIcon) {
        if (getActivity() != null) {
            AppCompatActivity activity = ((AppCompatActivity) getActivity());
            activity.setSupportActionBar(toolbar);

            boolean showBackIcon = getArguments() != null && getArguments().getBoolean(SHOW_BACK_ICON);
            if (showBackIcon) {
                if (activity.getSupportActionBar() != null) {
                    activity.getSupportActionBar().setDisplayHomeAsUpEnabled(true);
                    activity.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_navigation_back);
                    toolbar.setNavigationOnClickListener(v -> getActivity().finish());
                }
            }

            if (toolbarIcon != null) {
                toolbarIcon.setVisibility(showBackIcon ? View.GONE : View.VISIBLE);
            }
        }
    }

    protected <T> T setListener() {
        T listener;

        if (getParentFragment() != null) {
            listener = (T) getParentFragment();
        } else {
            try {
                listener = (T) getActivity();
            } catch (ClassCastException e) {
                listener = null;
            }
        }

        if (listener != null) {
            return listener;
        }

        throw new IllegalStateException("parent must implement listener");
    }


//    protected ApplicationComponent appComponent() {
//        return ((Nexx4App) getActivity().getApplication()).getAppComponent();
//    }
}

File: app/src/main/java/com/ShaHar91/poems/ui/base/BaseFragment.kt
package com.shahar91.poems.ui.base

import android.content.Intent
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.os.Bundle
import android.view.Menu
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.graphics.drawable.DrawableCompat
import androidx.fragment.app.Fragment
import com.shahar91.poems.MyApp
import com.shahar91.poems.R
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

abstract class BaseFragment<VM : BaseGoogleViewModel> :
    Fragment() {

    protected lateinit var viewModel: VM
    private lateinit var parentActivity: AppCompatActivity
    private var compositeDisposable: CompositeDisposable = CompositeDisposable()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        parentActivity = requireActivity() as AppCompatActivity
    }

    override fun onDestroy() {
        compositeDisposable.dispose()
        super.onDestroy()
    }

    protected fun configureToolbar(toolbar: Toolbar, color: Int? = null,
        showTitle: Boolean = true) {
        parentActivity.setSupportActionBar(toolbar)
        val showBackIcon = arguments != null && arguments!!.getBoolean(
            SHOW_BACK_ICON)
        if (showBackIcon) {
            parentActivity.supportActionBar?.let {
                it.setDisplayHomeAsUpEnabled(true)
                it.setHomeAsUpIndicator(R.drawable.ic_navigation_back)
            }
            color?.let { toolbar.navigationIcon?.colorFilter = PorterDuffColorFilter(it, PorterDuff.Mode.SRC_IN) }
            if (!showTitle) {
                toolbar.title = null
            }
            toolbar.setNavigationOnClickListener { parentActivity.onBackPressed() }
        }
    }

    fun tintMenuIcons(menu: Menu, color: Int) {
        for (i in 0 until menu.size()) {
            val item = menu.getItem(i)
            val normalDrawable = item.icon
            if (normalDrawable != null) {
                val wrapDrawable = DrawableCompat.wrap(
                    normalDrawable)
                DrawableCompat.setTint(wrapDrawable, color)
                item.icon = wrapDrawable
                //            } else if (item.getActionView() != null && item.getActionView() instanceof MediaRouteButton) {
                //                // Google Cast Button
                //                CastUtils.setCastButtonColor(getContext(), (MediaRouteButton) item.getActionView(), color);
            }
        }
    }

    protected fun addDisposable(disposable: Disposable?) {
        compositeDisposable.add(disposable!!)
    }

    protected fun <T> setListener(): T {
        val listener: T? = if (parentFragment != null) {
            parentFragment as T?
        } else {
            try {
                activity as T?
            } catch (e: ClassCastException) {
                null
            }
        }
        if (listener != null) {
            return listener
        }
        throw IllegalStateException("parent must implement listener")
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        for (fragment in childFragmentManager.fragments) {
            fragment.onActivityResult(requestCode, resultCode, data)
        }
    }

    companion object {
        const val SHOW_BACK_ICON = "SHOW_BACK_ICON"
    }
}